File: app/api/routes/analytics.py
"""
Analytics endpoints — compliance KPIs and chart data.
"""

from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.routes.auth import _current_user
from app.db.database import get_db
from app.models.user import User
from app.services import analytics_service as svc

router = APIRouter(prefix="/analytics", tags=["analytics"])


def _industry(current_user: User, override: Optional[str]) -> Optional[str]:
    if current_user.role == "admin" and override:
        return override
    return current_user.org_id


@router.get("/summary")
def summary(
    industry_id: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(_current_user),
):
    return svc.dashboard_summary(db, _industry(current_user, industry_id))


@router.get("/customers/risk-breakdown")
def customer_risk(
    industry_id: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(_current_user),
):
    return svc.customer_risk_breakdown(db, _industry(current_user, industry_id))


@router.get("/customers/onboarding-trend")
def onboarding_trend(
    days: int = Query(30, ge=7, le=365),
    industry_id: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(_current_user),
):
    return svc.customer_onboarding_trend(db, days, _industry(current_user, industry_id))


@router.get("/transactions/volume-trend")
def volume_trend(
    days: int = Query(30, ge=7, le=365),
    industry_id: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(_current_user),
):
    return svc.transaction_volume_trend(db, days, _industry(current_user, industry_id))


@router.get("/transactions/flagged-stats")
def flagged_stats(
    industry_id: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(_current_user),
):
    return svc.flagged_transaction_stats(db, _industry(current_user, industry_id))


@router.get("/kyc/status-breakdown")
def kyc_status(
    industry_id: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(_current_user),
):
    return svc.kyc_status_breakdown(db, _industry(current_user, industry_id))


@router.get("/reports/stats")
def report_stats(
    industry_id: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(_current_user),
):
    return svc.report_stats(db, _industry(current_user, industry_id))


@router.get("/reports/submission-trend")
def submission_trend(
    days: int = Query(90, ge=7, le=365),
    industry_id: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(_current_user),
):
    return svc.report_submission_trend(db, days, _industry(current_user, industry_id))


@router.get("/audit/activity-trend")
def audit_trend(
    days: int = Query(30, ge=7, le=365),
    industry_id: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(_current_user),
):
    return svc.audit_activity_trend(db, days, _industry(current_user, industry_id))

File: web/lib/api/analytics.ts
// Typed analytics-resource client — second pilot for the shared
// lib/api/client.ts pattern (see customers.ts for the first). Covers
// exactly the analytics endpoints the frontend actually calls today
// (app/analytics/{executive,compliance,mlro,operations}/page.tsx) — not
// app/api/routes/analytics.py's full endpoint list (customer/risk-
// breakdown and audit/activity-trend aren't called from the frontend and
// are left out). Types mirror app/services/analytics_service.py's actual
// return dicts, which aren't backed by a Pydantic response_model on the
// route side — confirmed by reading the service functions directly.

import { apiGet } from './client'

function qs(params: Record<string, string | number | undefined>): string {
  const q = new URLSearchParams()
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined) q.set(k, String(v))
  }
  const s = q.toString()
  return s ? `?${s}` : ''
}

export interface TrendPoint {
  date: string
  count: number
}

export interface VolumeTrendPoint extends TrendPoint {
  volume: number
}

export interface FlaggedTransactionStats {
  total: number
  flagged: number
  flagged_pct: number
}

export interface KycStatusBreakdown {
  total: number
  by_status: Record<string, number>
}

export interface ReportStats {
  total: number
  by_status: Record<string, number>
  by_type: Record<string, number>
}

export interface DashboardSummary {
  customers: { total: number; by_risk: Record<string, number> }
  transactions: FlaggedTransactionStats
  kyc: KycStatusBreakdown
  reports: ReportStats
  alerts: {
    pending_kyc_reviews: number
    overdue_reports: number
    high_risk_customers: number
  }
}

export interface PendingCustomerReviews {
  overdue: number
  due_within_30_days: number
}

export interface OpenCaseStats {
  open_total: number
  by_status: Record<string, number>
  by_severity: Record<string, number>
  overdue: number
  smr_candidates: number
}

export interface TrainingStatusBreakdown {
  total: number
  by_status: Record<string, number>
  overdue: number
  due_within_30_days: number
  completion_pct: number
}

export interface GovernanceOverview {
  policy_reviews_due_30d: number
  policy_reviews_overdue: number
  control_tests_overdue: number
  controls_total: number
  open_findings_total: number
  open_findings_by_risk: Record<string, number>
}

/** GET /analytics/summary */
export function getAnalyticsSummary(industryId?: string): Promise<DashboardSummary> {
  return apiGet(`/api/v1/analytics/summary${qs({ industry_id: industryId })}`)
}

/** GET /analytics/customers/onboarding-trend */
export function getCustomerOnboardingTrend(
  days = 30,
  industryId?: string
): Promise<TrendPoint[]> {
  return apiGet(
    `/api/v1/analytics/customers/onboarding-trend${qs({ days, industry_id: industryId })}`
  )
}

/** GET /analytics/customers/pending-reviews */
export function getPendingCustomerReviews(
  industryId?: string
): Promise<PendingCustomerReviews> {
  return apiGet(`/api/v1/analytics/customers/pending-reviews${qs({ industry_id: industryId })}`)
}

/** GET /analytics/training/status-breakdown */
export function getTrainingStatusBreakdown(
  industryId?: string
): Promise<TrainingStatusBreakdown> {
  return apiGet(`/api/v1/analytics/training/status-breakdown${qs({ industry_id: industryId })}`)
}

/** GET /analytics/governance/overview */
export function getGovernanceOverview(industryId?: string): Promise<GovernanceOverview> {
  return apiGet(`/api/v1/analytics/governance/overview${qs({ industry_id: industryId })}`)
}

/** GET /analytics/cases/open-stats */
export function getOpenCaseStats(industryId?: string): Promise<OpenCaseStats> {
  return apiGet(`/api/v1/analytics/cases/open-stats${qs({ industry_id: industryId })}`)
}

/** GET /analytics/transactions/volume-trend */
export function getTransactionVolumeTrend(
  days = 30,
  industryId?: string
): Promise<VolumeTrendPoint[]> {
  return apiGet(
    `/api/v1/analytics/transactions/volume-trend${qs({ days, industry_id: industryId })}`
  )
}

/** GET /analytics/reports/submission-trend */
export function getReportSubmissionTrend(
  days = 90,
  industryId?: string
): Promise<TrendPoint[]> {
  return apiGet(
    `/api/v1/analytics/reports/submission-trend${qs({ days, industry_id: industryId })}`
  )
}

/** GET /analytics/transactions/flagged-stats */
export function getFlaggedTransactionStats(
  industryId?: string
): Promise<FlaggedTransactionStats> {
  return apiGet(`/api/v1/analytics/transactions/flagged-stats${qs({ industry_id: industryId })}`)
}

/** GET /analytics/reports/stats */
export function getReportStats(industryId?: string): Promise<ReportStats> {
  return apiGet(`/api/v1/analytics/reports/stats${qs({ industry_id: industryId })}`)
}

/** GET /analytics/kyc/status-breakdown */
export function getKycStatusBreakdown(industryId?: string): Promise<KycStatusBreakdown> {
  return apiGet(`/api/v1/analytics/kyc/status-breakdown${qs({ industry_id: industryId })}`)
}
